_posts/2019-04-06-trees-and-rectangles-reconciling-sql-and-recursive-programming-languages.md

```diff
@@ -0,0 +1,159 @@
+
+How do we resolve this?
+
+### Object-Relational Mappers (or: teaching trees to think about rectangles)
+
+ORMs were one stab at this problem. The underlying idea was that you
+could teach the object system about the structure of a relational
+database, and then make database queries lazily when the pattern of
+calls required the data, thereby letting the developer stay primarily
+in a hierarchical pattern of thought.
+
+While this does sound appealing, it turns out to have many problems in
+practice, from the
+[n+1 query problem](https://medium.com/@bretdoucette/n-1-queries-and-how-to-avoid-them-a12f02345be5)
+to nasty interactions between data cached in objects and the underlying database state. I won't go
+into this too deeply, as [Ted Neward](http://blogs.tedneward.com/post/the-vietnam-of-computer-science/)
+and [Mark Rickerby](https://maetl.net/talks/rise-and-fall-of-orm) have
+already covered it admirably, but in essence: it isn't feasible to
+ignore the structures and guarantees of the underlying relational
+database if you want either correctness or performance. Rails still
+uses this pattern, but even there the Arel package is used to allow
+devs to talk about queries in a way that maps closely to the
+database.
+
+
+### NoSQL! (kill all the rectangles, talk trees)
+
+To massively oversimplify, the term "NoSQL" became popular around 2009
+to describe a disparate group of databases that did not rely primarily
+on the relational view of data as rectangles. Arguably, the earlier object
+database approach was a similar idea, the idea being in least at part
+that if you couldn't easily map tree-based structures to rectangular
+tables, why not get rid of the rectangular data store entirely? (Graph
+databases aren't that far from this approach either.)
+
+This isn't inherently a broken model. As long as the whole object
+graph is maintained, it can even be faster - traversals can be done
+via pointers rather than by consulting indexes on keys. (It also has
+to be said that SQL appears to be difficult to scale to enormous
+datasets, prompting the development of stores like BigTable.)
+
+Frequently it's modelled in unfortunate ways, however. Sarah Mei's
+[breakdown](http://www.sarahmei.com/blog/2013/11/11/why-you-should-never-use-mongodb/)
+of why not to use Mongo is a good description of what can happen when
+you serialise object data as text, rather than preserving the graph as
+a whole.
+
+Even assuming that you manage to avoid that particular pitfall,
+however, if query patterns change you could still be in trouble. One
+of SQL's big advantages is that it supports ad-hoc, unexpected
+queries, and that the indices can be tweaked without changing anything
+about the structure of the store itself.
+
+### kill all the trees, talk rectangles
+
+Logically, if you look at the 2x2, there should be a way to talk
+rectangles all the way through. I have no idea at all what this would
+look like. APL on the front end?
```

Comment:
https://github.com/tonsky/datascript, perhaps?  I've never read it hard enough to know for sure.  There's some relationship to the delightfully named (though probably not germane) https://mozilla.github.io/mentat/, I think?

I'm pretty sure APL rectangles aren't the same kind as RDB rectangles...